Lib/pickletools.py

```diff
@@ -2880,11 +2880,14 @@ def _test():
         if not args.pickle_file:
             parser.print_help()
         elif len(args.pickle_file) == 1:
-            dis(args.pickle_file[0], args.output, None,
-                args.indentlevel, annotate)
+            with args.pickle_file[0], args.output:
+                dis(args.pickle_file[0], args.output, None,
+                    args.indentlevel, annotate)
         else:
             memo = {} if args.memo else None
-            for f in args.pickle_file:
-                preamble = args.preamble.format(name=f.name)
-                args.output.write(preamble + '\n')
-                dis(f, args.output, memo, args.indentlevel, annotate)
+            with args.output:
+                for f in args.pickle_file:
+                    with f:
```

Comment:
I think this could be improved as if there is an error while processing some files, the files that have not been already processed will not be closed. By using `contextlib.ExitStack()` here we could make sure that the files would always be closed even if some error occurs as we would register them all before starting processing them.

Reply:
Sounds good!
Just for making sure, you mean sth like this?

```
with contextlib.ExitStack() as stack:                            
    output = stack.enter_context(args.output)                    
    infiles = [stack.enter_context(f) for f in args.pickle_file] 
    for f in infiles:                                            
        preamble = args.preamble.format(name=f.name)
        output.write(preamble + '\n')
        dis(f, output, memo, args.indentlevel, annotate)
```

Reply:
What about just using `try/finally`? Looks simpler to me:

```
            try:
                for f in args.pickle_file:
                    preamble = args.preamble.format(name=f.name)
                    args.output.write(preamble + '\n')
                    dis(f, args.output, memo, args.indentlevel, annotate)
            finally:
                args.output.close()
                for f in args.pickle_file:
                    f.close()
```